Mathlib/Tactic/Simp/FinsetInterval.lean

```diff
@@ -0,0 +1,140 @@
+    | return .continue
+  let hm : Q(IsNat (OfNat.ofNat $em) $em) := q(⟨rfl⟩)
+  let hn : Q(IsNat (OfNat.ofNat $en) $en) := q(⟨rfl⟩)
+  unless em.isRawNatLit && en.isRawNatLit do
+    return .continue
+  let m := em.natLit!
+  let n := en.natLit!
+  let ⟨s, p⟩ ← evalFinsetIccNat m n hm hn
+  return .done { expr := s, proof? := p }
+
+/-- Simproc to compute `Finset.Ico a b` when `a b : ℕ`. -/
+simproc_decl Ico_nat (Ico _ _) := fun e ↦ do
+  let ⟨1, ~q(Finset ℕ), ~q(Ico (OfNat.ofNat $em) (OfNat.ofNat $en))⟩ ← inferTypeQ e
+    | return .continue
+  let hm : Q(IsNat (OfNat.ofNat $em) $em) := q(⟨rfl⟩)
+  let hn : Q(IsNat (OfNat.ofNat $en) $en) := q(⟨rfl⟩)
+  unless em.isRawNatLit && en.isRawNatLit do
+    return .continue
+  let m := em.natLit!
+  let n := en.natLit!
+  match n with
+  | 0 =>
+    return .done { expr := (q(∅) : Q(Finset ℕ)), proof? := q(Ico_zero $em) }
+  | n + 1 =>
+    let hn₀ ← mkDecideProofQq q($en ≠ 0)
+    let hn := q(isNat_natPred $hn rfl)
+    let ⟨s, p⟩ ← evalFinsetIccNat m n hm hn
+    return .done { expr := s, proof? := q(Ico_eq_of_Icc_pred_eq $hn₀ $p) }
+
+/-- Simproc to compute `Finset.Ioc a b` when `a b : ℕ`. -/
+simproc_decl Ioc_nat (Ioc _ _) := fun e ↦ do
+  let ⟨1, ~q(Finset ℕ), ~q(Ioc (OfNat.ofNat $em) (OfNat.ofNat $en))⟩ ← inferTypeQ e
+    | return .continue
+  let hm : Q(IsNat (OfNat.ofNat $em) $em) := q(⟨rfl⟩)
+  let hn : Q(IsNat (OfNat.ofNat $en) $en) := q(⟨rfl⟩)
+  unless em.isRawNatLit && en.isRawNatLit do
+    return .continue
+  let m := em.natLit!
+  let n := en.natLit!
+  let hm := q(isNat_natSucc $hm rfl)
+  let ⟨s, p⟩ ← evalFinsetIccNat (m + 1) n hm hn
+  return .done { expr := s, proof? := q(Ioc_eq_of_Icc_succ_eq $p) }
+
+/-- Simproc to compute `Finset.Ioo a b` when `a b : ℕ`. -/
+simproc_decl Ioo_nat (Ioo _ _) := fun e ↦ do
+  let ⟨1, ~q(Finset ℕ), ~q(Ioo (OfNat.ofNat $em) (OfNat.ofNat $en))⟩ ← inferTypeQ e
+    | return .continue
+  let hm : Q(IsNat (OfNat.ofNat $em) $em) := q(⟨rfl⟩)
+  let hn : Q(IsNat (OfNat.ofNat $en) $en) := q(⟨rfl⟩)
+  unless em.isRawNatLit && en.isRawNatLit do
+    return .continue
+  let m := em.natLit!
+  let n := en.natLit!
+  let hm := q(isNat_natSucc $hm rfl)
+  let hn := q(isNat_natPred $hn rfl)
+  let ⟨s, p⟩ ← evalFinsetIccNat (m + 1) (n - 1) hm hn
+  return .done { expr := s, proof? := q(Ioo_eq_of_Icc_succ_pred_eq $p) }
+
+example : Icc 1 0 = ∅ := by simp only [Icc_nat]
+example : Icc 1 1 = {1} := by simp only [Icc_nat]
+example : Icc 1 2 = {1, 2} := by simp only [Icc_nat]
+
+example : Ico 1 1 = ∅ := by simp only [Ico_nat]
+example : Ico 1 2 = {1} := by simp only [Ico_nat]
+example : Ico 1 3 = {1, 2} := by simp only [Ico_nat]
+
+example : Ioc 1 1 = ∅ := by simp only [Ioc_nat]
+example : Ioc 1 2 = {2} := by simp only [Ioc_nat]
+example : Ioc 1 3 = {2, 3} := by simp only [Ioc_nat]
+
+example : Ioo 1 2 = ∅ := by simp only [Ioo_nat]
+example : Ioo 1 3 = {2} := by simp only [Ioo_nat]
+example : Ioo 1 4 = {2, 3} := by simp only [Ioo_nat]
```

Comment:
Can you add some tests of reversed bounds?

Reply:
... as in `Icc a b` where `b < a`? Isn't the first one enough?